[3.x] Allow rendering of 3d content at lower resolution

Ansraer opened the pull request:
This makes it possible to dynamically adjust the rendering resolution of 3D content. 2D content and UI stuff are still rendered at full resolution.

The user can adjust the target resolution with the `Rendering/Quality/3D/Lower Resolution` project setting. 

The editor displaying a scene at 1/4 the resolution
![image](https://user-images.githubusercontent.com/6329420/225271236-becc112c-415c-421b-b823-7e9942e9b357.png)


---
This PR was sponsored by Ramatak with 💚

Ansraer committed:
allow rendering of 3d content at lower resolution

fire commented:
This is useful for mobile devices that require more performance.

Calinou commented:
Consider adding an option to toggle between nearest-neighbor and bilinear filtering, as nearest-neighbor filtering is noticeably sharper. Pixels will have even scaling if using a scale factor of `0.25`, `0.333` or `0.5`.

clayjohn commented:
I have a feeling that this should be exposed on a per-viewport basis as it is in 4.0. Doing viewport within viewport is pretty common for things like screens and gun scopes. Users will need the option to render at their selected resolution for the inner Viewports or else it will be very difficult to manage

Ansraer commented:
Updated the PR. The settings option has been renamed and an additional option for the filter mode has been added.

I also added overwrites for specific viewports to the Viewport Node:

![image](https://user-images.githubusercontent.com/6329420/226921924-68830209-a0aa-480f-b90e-ecd9cc0c7892.png)

> If it works correctly, please allow resolution scales up to 2.0 to be used for supersampling like in 4.x. This provides SSAA which is demanding, but is effective on specular aliasing (there's no TAA in 3.x).

This won't be possible with my current code. At Ramatak our goal is to dynamically change the resolution while the game is running, so this PR doesn't actually change the size of the render buffer (buffer allocations are expensive).

Calinou commented:
> This won't be possible with my current code. At Ramatak our goal is to dynamically change the resolution while the game is running, so this PR doesn't actually change the size of the render buffer (buffer allocations are expensive).

I see. We should try to support this [use case](https://github.com/godotengine/godot-proposals/issues/917) in 4.x as well, likely by adding a separate `scaling_3d_max` that defines the actual render buffer size (with `scaling_3d_scale` being adjustable without recreating render buffers if below `scaling_3d_max`).

clayjohn reviewed (commented):
I think this should work more like it does in 4.0. Primarily in 2 ways:
1. The upscaling should happen at the time you copy the 3D buffers into the 2D buffer i.e. during the final post-process pass, not before post processing occurs. The reason is twofold 1) performance, you avoid a couple full screen copies and you can avoid doing Glow, auto-exposure, DoF at full res 2) quality, things like glow and DoF will build on the upscaling artifacts, plus they are low frequency effects so they will suffer from upscaling artifacts less. 

Also, since you are doing a full screen blit anyway, you should have no problem supporting MSAA. When rendering at a lower resolution AA becomes much more important as the upscaling will make AA artifacts even worse.
